api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -172,42 +182,50 @@ public void close() {
 
   private void wakeUp(Instant now) {
     LOGGER.debug("Waking up...");
-    Duration delay = nextTokenRefresh(getCurrentTokens(), now, Duration.ZERO);
+    Tokens currentTokens = getCurrentTokensIfAvailable();
+    Duration delay = nextTokenRefresh(currentTokens, now, Duration.ZERO);
     if (delay.compareTo(MIN_REFRESH_DELAY) < 0) {
       LOGGER.debug("Refreshing tokens immediately");
       renewTokens();
     } else {
+      LOGGER.debug("Tokens are still valid");
       scheduleTokensRenewal(delay);
     }
   }
 
-  private Tokens maybeScheduleTokensRenewal(Tokens currentTokens) {
+  private void maybeScheduleTokensRenewal(Tokens currentTokens) {
     Instant now = clock.get();
     if (Duration.between(lastAccess, now).compareTo(idleInterval) > 0) {
       sleeping.set(true);
       LOGGER.debug("Sleeping...");
     } else {
-      scheduleTokensRenewal(nextTokenRefresh(currentTokens, now, MIN_REFRESH_DELAY));
+      Duration delay = nextTokenRefresh(currentTokens, now, MIN_REFRESH_DELAY);
+      scheduleTokensRenewal(delay);
     }
-    return currentTokens;
   }
 
   private void scheduleTokensRenewal(Duration delay) {
     LOGGER.debug("Scheduling token refresh in {}", delay);
-    tokenRefreshFuture =
-        executor.schedule(this::renewTokens, delay.toMillis(), TimeUnit.MILLISECONDS);
+    try {
+      tokenRefreshFuture =
+          executor.schedule(this::renewTokens, delay.toMillis(), TimeUnit.MILLISECONDS);
+    } catch (RejectedExecutionException e) {
+      LOGGER.warn("Failed to schedule next token renewal, forcibly sleeping", e);
+      sleeping.set(true);
+    }
   }
 
   private void renewTokens() {
     CompletionStage<Tokens> oldTokensStage = currentTokensStage;
     currentTokensStage =
         oldTokensStage
-            // try refreshing the current tokens
+            // try refreshing the current tokens (if they exist)
             .thenApply(this::refreshTokens)
```

Comment:
Not really related to this change, but I did not notice this before... Will this create a chain of stages that grows indefinitely with every `renewTokens` call?

Reply:
I guess it's fine, because `renewTokens()` will be called only after `currentTokensStage` completes.

Reply:
No: each (incomplete) future holds references to its direct children, but not to its parent. Furthermore, when a future completes, the completion signal is passed downstream to all its children, then all child references are cleared.